freqtrade/exchange/binance_public_data.py

```diff
@@ -0,0 +1,210 @@
+"""
+Fetch daily-archived OHLCV data from https://data.binance.vision/
+"""
+
+import asyncio
+import datetime
+import io
+import itertools
+import logging
+import zipfile
+
+import aiohttp
+import pandas as pd
+from pandas import DataFrame
+
+from freqtrade.enums import CandleType
+from freqtrade.util.datetime_helpers import dt_from_ts, dt_now
+
+
+logger = logging.getLogger(__name__)
+
+
+class BadHttpStatus(Exception):
+    """Not 200/404"""
+
+    pass
+
+
+async def fetch_ohlcv(
+    candle_type: CandleType,
+    pair: str,
+    timeframe: str,
+    since_ms: int,
+    until_ms: int | None,
+    stop_on_404: bool = False,
+) -> DataFrame:
+    """
+    Fetch OHLCV data from https://data.binance.vision/
+    :candle_type: Currently only spot and futures are supported
+    :param until_ms: `None` indicates the timestamp of the latest available data
+    :param stop_on_404: Stop to download the following data when a 404 returned
+    :return: the date range is between [since_ms, until_ms),
+        return None if no data available in the time range
+    """
+    if candle_type == CandleType.SPOT:
+        asset_type = "spot"
+    elif candle_type == CandleType.FUTURES:
+        asset_type = "futures/um"
+    else:
+        raise ValueError(f"Unsupported CandleType: {candle_type}")
+    symbol = symbol_ccxt_to_binance(pair)
+    start = dt_from_ts(since_ms)
+    end = dt_from_ts(until_ms) if until_ms else dt_now()
+
+    # We use two days ago as the last available day because the daily archives are daily uploaded
+    # and have several hours delay
+    last_available_date = dt_now() - datetime.timedelta(days=2)
+    end = min(end, last_available_date)
+    if start >= end:
+        return DataFrame()
+    df = await _fetch_ohlcv(asset_type, symbol, timeframe, start, end, stop_on_404)
+    logger.info(
+        f"Downloaded data for {pair} from https://data.binance.vision/ with length {len(df)}."
+    )
+    if not df.empty:
+        return df.loc[(df["date"] >= start) & (df["date"] < end)]
+    else:
+        return df
+
+
+def symbol_ccxt_to_binance(symbol: str) -> str:
+    """
+    Convert ccxt symbol notation to binance notation
+    e.g. BTC/USDT -> BTCUSDT, BTC/USDT:USDT -> BTCUSDT
+    """
+    if ":" in symbol:
+        parts = symbol.split(":")
+        if len(parts) != 2:
+            raise ValueError(f"Cannot recognize symbol: {symbol}")
+        return parts[0].replace("/", "")
+    else:
+        return symbol.replace("/", "")
+
```

Comment:
I'm not sure that's necessarily correct for all cases - better rely on the ccxt markets object - it contains the original symbol.

Reply:
The reason why I'm firstly reluctant to use ccxt is that exchangeInfo endpoint consumes a lot of time (maybe 2 seconds) and I need to patch this function or ccxt in the unittest.

I have checked all USD-M futures and I can confirm that all of the **perpetual contracts** (but not include dilivery) conform to this pattern: `COIN/USDT:USDT` or `COIN/USDC:USDC` (although ccxt and binance may use different COIN names)

Reply:
> (although ccxt and binance may use different COIN names)

wouldn't this mean that ccxt uses BTC/USDT:USDT - but on bybit it's XBT_USDT ... ? which would mean the logic above will fail.
Best example is `YOYOW/BTC` - which is actually `YOYOBTC` (missing a W) at the exchange level - so i don't think you can get around using the ccxt market object.

While it'll add 2s to the overall time - it'll also improve reliability - and you can really pass that in as object (`exchange.markets` - so you don't need the exchange object itself).

Reply:
Agreed we should use CCXT.

Reply:
[refactor: use CCXT for pair to symbol conversion](https://github.com/freqtrade/freqtrade/pull/10858/commits/37726fba58834e922bc8079141aa57d5585f08ec)